adapters/akka-http/src/main/scala/caliban/AkkaHttpAdapter.scala

```diff
@@ -149,96 +149,104 @@ trait AkkaHttpAdapter {
-                  IO.fromFuture(_ => sendTo.offer(TextMessage(s"""{"type":"complete","id":"$messageId"}""")))
-            }
-      } yield ()
-
     get {
-      extractUpgradeToWebSocket { upgrade =>
-        val (queue, source) = Source.queue[Message](0, OverflowStrategy.fail).preMaterialize()
-        val subscriptions   = runtime.unsafeRun(Ref.make(Map.empty[Option[String], Fiber[Throwable, Unit]]))
-        val sink = Sink.foreach[Message] {
-          case TextMessage.Strict(text) =>
-            val io = for {
-              msg     <- Task.fromEither(json.parseWSMessage(text))
-              msgType = msg.messageType
-              _ <- RIO.whenCase(msgType) {
-                    case "connection_init" =>
-                      Task.fromFuture(_ => queue.offer(TextMessage("""{"type":"connection_ack"}"""))) *>
-                        Task.whenCase(keepAliveTime) {
-                          case Some(time) =>
-                            // Save the keep-alive fiber with a key of None so that it's interrupted later
-                            IO.fromFuture(_ => queue.offer(TextMessage("""{"type":"ka"}""")))
-                              .repeat(Schedule.spaced(time))
-                              .provideLayer(Clock.live)
-                              .unit
-                              .forkDaemon
-                              .flatMap(keepAliveFiber => subscriptions.update(_.updated(None, keepAliveFiber)))
-                        }
-                    case "connection_terminate" =>
-                      IO.effect(queue.complete())
-                    case "start" =>
-                      RIO.whenCase(msg.request) {
-                        case Some(req) =>
-                          startSubscription(msg.id, req, queue, subscriptions).catchAll(error =>
-                            IO.fromFuture(_ => queue.offer(TextMessage(json.encodeWSError(msg.id, error))))
-                          )
-                      }
-                    case "stop" =>
-                      subscriptions
-                        .modify(map => (map.get(Option(msg.id)), map - Option(msg.id)))
-                        .flatMap(fiber =>
-                          IO.whenCase(fiber) {
-                            case Some(fiber) =>
-                              fiber.interrupt *>
-                                IO.fromFuture(_ =>
-                                  queue.offer(TextMessage(s"""{"type":"complete","id":"${msg.id}"}"""))
-                                )
-                          }
+      extractRequestContext { ctx =>
+        extractUpgradeToWebSocket { upgrade =>
+          def sendMessage(
+            sendQueue: SourceQueueWithComplete[Message],
+            id: String,
+            data: ResponseValue,
+            errors: List[E]
+          ): Task[QueueOfferResult] =
+            IO.fromFuture(_ => sendQueue.offer(TextMessage(json.encodeWSResponse(id, data, errors))))
+
+          def startSubscription(
```

Comment:
How about just adding `ctx` as a parameter to this function and keep it outside as it was? Just to reduce allocations done at each request and keep it slightly more readable.